File: junk/llvmtest.php
<?php

$llvm = FFI::cdef(
    <<<EOT
    typedef struct LLVMOpaqueTarget *LLVMTargetRef;

    char *LLVMGetDefaultTargetTriple(void);
    int LLVMGetTargetFromTriple(const char *Triple, LLVMTargetRef *OutTarget, char **ErrorMessage);
    char *LLVMGetTargetName(LLVMTargetRef T);
    char *LLVMGetTargetDescription(LLVMTargetRef T);
    void LLVMDisposeMessage(char *Message);

    LLVMInitializeAllTargetInfos();
    LLVMInitializeAllTargetMCs();
    LLVMInitializeAllAsmParsers();
    LLVMInitializeAllAsmPrinters();

    void LLVMInitializeAArch64Target(void);
    void LLVMInitializeAArch64TargetMC(void);
    void LLVMInitializeAArch64AsmPrinter(void);
    void LLVMInitializeAArch64AsmParser(void);

    EOT,
    "/opt/homebrew/Cellar/llvm/19.1.6/lib/libLLVM-C.dylib" // Load the LLVM shared library
);

// Initialize AArch64 targets
$llvm->LLVMInitializeAArch64Target();
$llvm->LLVMInitializeAArch64TargetMC();
$llvm->LLVMInitializeAArch64AsmPrinter();
$llvm->LLVMInitializeAArch64AsmParser();

$llvm->LLVMInitializeAllTargetInfos();
$llvm->LLVMInitializeAllTargetMCs();
$llvm->LLVMInitializeAllAsmParsers();
$llvm->LLVMInitializeAllAsmPrinters();


// Set the target triple to AArch64
$triple = "aarch64-apple-darwin";
echo "Using target triple: $triple\n";

$targetPtr = $llvm->new("LLVMTargetRef*");
$errorMessage = $llvm->new("char*");

if ($llvm->LLVMGetTargetFromTriple($triple, FFI::addr($targetPtr), FFI::addr($errorMessage))) {
    // Handle errors
    echo "Error: " . FFI::string($errorMessage) . "\n";
    $llvm->LLVMDisposeMessage($errorMessage);
    exit(1);
}

File: junk/llvmtest.c
#include <llvm-c/Target.h>
#include <llvm-c/TargetMachine.h>
#include <stdio.h>

// clang -I/opt/homebrew/Cellar/llvm/19.1.6/include -L/opt/homebrew/Cellar/llvm/19.1.6/lib -lLLVM -o test_llvm llvmtest.c

int main() {
    // Initialize AArch64 target and related components
    // printf("Initializing AArch64 target...\n");
    // LLVMInitializeAArch64Target();
    // LLVMInitializeAArch64TargetMC();
    // LLVMInitializeAArch64AsmPrinter();
    // LLVMInitializeAArch64AsmParser();

    // Initialize all targets
    printf("Initializing all targets...\n");
    LLVMInitializeAllTargetInfos();
    LLVMInitializeAllTargetMCs();
    LLVMInitializeAllAsmParsers();
    LLVMInitializeAllAsmPrinters();

    // Manually set the triple to AArch64
    const char *triple = "aarch64-apple-darwin";  // Adjust as needed (this is macOS AArch64)
    printf("Using target triple: %s\n", triple);

    // Retrieve the target using the triple
    LLVMTargetRef target = NULL;
    char *errorMessage = NULL;
    if (LLVMGetTargetFromTriple(triple, &target, &errorMessage) != 0) {
        printf("Error: %s\n", errorMessage);
        //LLVMDisposeMessage(errorMessage);
        return 1;
    }

    // Print target details
    const char *targetName = LLVMGetTargetName(target);
    const char *targetDesc = LLVMGetTargetDescription(target);

    printf("Target Name: %s\n", targetName);
    printf("Target Description: %s\n", targetDesc);

    return 0;
}

